pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+// NodeList is only nil in testing.
+func NewMultiCIDRRangeAllocator(
+	client clientset.Interface,
+	nodeInformer informers.NodeInformer,
+	clusterCIDRInformer networkinginformers.ClusterCIDRInformer,
+	allocatorParams CIDRAllocatorParams,
+	nodeList *v1.NodeList,
+	testCIDRMap map[string][]*cidrset.ClusterCIDR,
+) (CIDRAllocator, error) {
+	if client == nil {
+		klog.Fatalf("client is nil")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventSource := v1.EventSource{
+		Component: "multiCIDRRangeAllocator",
+	}
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, eventSource)
+	eventBroadcaster.StartStructuredLogging(0)
+	klog.V(0).Infof("Started sending events to API Server. (EventSource = %v)", eventSource)
+
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: client.CoreV1().Events(""),
+		})
+
+	if client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		ratelimiter.RegisterMetricAndTrackRateLimiterUsage("multi_cidr_range_allocator", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	ra := &multiCIDRRangeAllocator{
+		client:                client,
+		nodeLister:            nodeInformer.Lister(),
+		nodesSynced:           nodeInformer.Informer().HasSynced,
+		clusterCIDRLister:     clusterCIDRInformer.Lister(),
+		clusterCIDRSynced:     clusterCIDRInformer.Informer().HasSynced,
+		nodeCIDRUpdateChannel: make(chan multiCIDRNodeReservedCIDRs, cidrUpdateQueueSize),
+		recorder:              recorder,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "multi_cidr_range_allocator"),
+		lock:                  &sync.Mutex{},
+		nodesInProcessing:     sets.NewString(),
+		cidrMap:               make(map[string][]*cidrset.ClusterCIDR, 0),
+	}
+
+	// testCIDRMap is only set for testing purposes.
+	if len(testCIDRMap) > 0 {
+		ra.cidrMap = testCIDRMap
+		klog.Warningf("testCIDRMap should only be set for testing purposes, if this is seen in production logs, it might be a misconfiguration or a bug.")
+	}
+
+	ccList, err := listClusterCIDRs(client)
+	if err != nil {
+		return nil, err
+	}
+
+	if ccList == nil {
+		ccList = &networkingv1alpha1.ClusterCIDRList{}
+	}
+	createDefaultClusterCIDR(ccList, allocatorParams)
+
+	// Regenerate the cidrMaps from the existing ClusterCIDRs.
+	for _, clusterCIDR := range ccList.Items {
+		klog.Infof("Regenerating existing ClusterCIDR: %v", clusterCIDR)
+		// Create an event for invalid ClusterCIDRs, do not crash on failures.
+		if err := ra.reconcileBootstrap(&clusterCIDR); err != nil {
+			klog.Errorf("Error while regenerating existing ClusterCIDR: %v", err)
+			ra.recorder.Event(&clusterCIDR, "Warning", "InvalidClusterCIDR encountered while regenerating ClusterCIDR during bootstrap.", err.Error())
+		}
+	}
+
+	clusterCIDRInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    createClusterCIDRHandler(ra.reconcileCreate),
```

Comment:
don't we should handle updates, what happen if an user sets or removes our finalizer?

Reply:
I am not 100% sure, but I guess the metadata is handled by some other controller, and we do not need to worry about that here, we are just checking the presence of the finalizer

Reply:
The thing is, imagine that I do a kubectk edit on the resources as power user and remove the finalizer